### PR TITLE
allow id generators to be passed to the tracer (#209)

### PR DESCRIPTION
**Pull request by rcoedo**
This PR tries to address the issue #209.

The tracer now defines two extra optional function parameters, the traceId generator function and the spanId generator function. The spanId generator parameter defaults to the `randomTraceId` function that we were already using. The traceId generator parameter defaults to a new `randomTraceId128bit` function that mimics the current behaviour: `randomTraceId() + randomTraceId()`.

I am not sure this is the best approach but it is the best solution I could find without breaking the API or changing the current behaviour. I lack the deep understanding of zipkin to feel comfortable making deeper changes, so any suggestions will be welcomed 😄

cheers!

### CODE REVIEW

**Comment by jcchavezs** [comment]
This looks legit to me, tho I would like to see how you did improve the traceID randomness and maybe add it as the standard way to create IDs? Ping @adriancole @ghermeto

**Comment by rcoedo** [comment]
I think the randomness would improved by using something like `uuid` to generate the ids, but it would probably affect performance. With the current approach a developer would be able to choose their own algorithm if needed.

If you think the current default id generation is not good enough, I could default to `uuids` instead of the `randomTraceId` function.

**Comment by jcchavezs** [comment]
I see. I though you tweaked the existing randomTraceId. It makes sense we
make it pluggable. I think UUID is great but for not too high load you
might not need that much randomness.

José Carlos Chávez


Den tor. 21. feb. 2019 kl. 10:19 skrev Roman Coedo <notifications@github.com

**Comment by adriancole** [comment]
my opinion is that we should vet that random isnt working properly and maybe cite why. the issue leading to this sounded a bit like a hypothesis and we probably shouldn't add new APIs based on that rationale.

sometimes people think id collision is happening when really it is resource leak (like a TLS thing)

it is possible but unlikely that the actual problem is that two 64 bit numbers are clashing in a short amount of time.

anyway tldr is please consider not changing public api or making more knobs until the problem is vetted and something like this . I would also suggest rule of 3. like one other high volume site like netflix

**Comment by adriancole** [comment]
sorry meant CLS not TLS (like id leaking on thread which is a frequent root cause when folks start pointing fingers at RNG because it has the same effect)

**Comment by adriancole** [comment]
ps I was mentioned and so did a "drive by" suggestion. I'm feeling under the weather so very likely not to continue this thread in the next day.. iotw dont block on me  just consider and decide how to proceed!

**Comment by ghermeto** [comment]
@adriancole I will ask Insights if they have instrumentation to check for collisions, but I personally haven't seen it so far.

**Comment by jcchavezs** [comment]
After a second thought I also believe it is very unlike that we have collisions in generated traces and it looks to me more what Adrian says about CLS. I will close this issue unless @rcoedo is terribly sure this is the case (maybe with a high load test).
